File: 0463-island-perimeter/0463-island-perimeter.py
class Solution:
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        row = len(grid)
        col= len(grid[0])
        direc = [(0,1),(1,0),(0,-1),(-1,0)]
        vis = set()
        def bfs(r, c):
            peri = 0
            qu = deque()
            qu.append((r, c))
            vis.add((r,c)) 
            while qu:
                a, b = qu.popleft()
                for dx, dy in direc:
                    nx = dx + a
                    ny = dy + b
                    if(nx < 0 or ny < 0  or nx >= row or ny >= col or grid[nx][ny] == 0):
                        peri += 1
                    elif (nx, ny) not in vis:
                        vis.add((nx, ny))
                        qu.append((nx, ny))
            return peri
        for i in range(row):
            for j in range(col):
                if grid[i][j] == 1:
                    return bfs(i, j)
        return 0  

File: 0463-island-perimeter/0463-island-perimeter.cpp
class Solution {
public:
    int islandPerimeter(vector<vector<int>>& grid) {
        int row = grid.size(), col = grid[0].size();
        vector<vector<bool>> vis(row,vector<bool>(col, false));
        int dir[4][2] = {{0,1},{1,0},{0,-1},{-1,0}};
        for(int i = 0; i < row; i++){
            for(int j = 0; j < col; j++){
                if(grid[i][j] == 1){
                    queue<pair<int, int>> qu;
                    qu.push({i, j});
                    vis[i][j] = true;
                    int peri =0 ;
                    while(!qu.empty()){
                        auto[a, b] = qu.front();
                        qu.pop();
                        for(auto& di: dir){
                            int nx = a + di[0], ny = b + di[1];
                            if(nx < 0 || nx >= row || ny < 0 || ny >= col || grid[nx][ny] == 0)peri++;
                            else if(vis[nx][ny] == false){
                                vis[nx][ny] = true;
                                qu.push({nx, ny});
                            }

                        }
                    }
                    return peri;
                }
            }
        }
        return 0;
    }
};